src/devices/bus/rs232/swtpc8212.cpp

```diff
@@ -0,0 +1,74 @@
+// license:BSD-3-Clause
+// copyright-holders:68bit
+
+#include "emu.h"
+#include "swtpc8212.h"
+
+swtpc8212_terminal_device::swtpc8212_terminal_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, SERIAL_TERMINAL_SWTPC8212, tag, owner, clock)
+	, device_rs232_port_interface(mconfig, *this)
+	, m_swtpc8212(*this, "swtpc8212")
+	, m_flow_control(*this, "flow_control")
+{
+}
+
+void swtpc8212_terminal_device::device_add_mconfig(machine_config &config)
+{
+	SWTPC8212(config, m_swtpc8212, 0);
+	m_swtpc8212->rs232_conn_txd_handler().set(FUNC(swtpc8212_terminal_device::output_rxd));
+	m_swtpc8212->rs232_conn_rts_handler().set(FUNC(swtpc8212_terminal_device::route_term_rts));
+	m_swtpc8212->rs232_conn_dtr_handler().set(FUNC(swtpc8212_terminal_device::route_term_dtr));
+}
+
+INPUT_PORTS_START(swtpc8212_terminal)
+
+	PORT_START("flow_control")
+	PORT_CONFNAME(0xf, 1, "Flow control")
+	PORT_CONFSETTING(0x00, "None")
+	PORT_CONFSETTING(0x01, "Terminal DTR to remote CTS")
+
+INPUT_PORTS_END
+
+ioport_constructor swtpc8212_terminal_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(swtpc8212_terminal);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_terminal_device::input_txd)
+{
+	m_swtpc8212->rs232_conn_rxd_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_terminal_device::route_term_rts)
+{
+	// Loop the terminal RTS output to the terminal CTS input.
+	m_swtpc8212->rs232_conn_cts_w(state);
+}
+
+// This terminal uses DTR for hardware flow control.
+WRITE_LINE_MEMBER(swtpc8212_terminal_device::route_term_dtr)
+{
+	if (m_flow_control->read() == 1)
+	{
+		// Connect the terminal DTR output to CTS at the other end.
+		swtpc8212_terminal_device::output_cts(state);
+	}
+}
+
+void swtpc8212_terminal_device::device_start()
+{
+}
+
+void swtpc8212_terminal_device::device_reset()
+{
+	// To the terminal
+	m_swtpc8212->rs232_conn_cts_w(0);
+
+	// To the computer
+	output_rxd(1);
+	output_dcd(0);
+	output_dsr(0);
+	output_cts(0);
```

Comment:
You shouldn't be doing this in `device_reset` – if the terminal actually changes line state on reset it should push the changes out itself.  Doing it in this shim device could cause the state seen by the other side of the port to get out of sync.

Reply:
This layer, the rs232 bus, is more than a straight-through connection of the terminal to the host, rather it actively directs and controls the lines, it is one of many forms of a null-modem cable, and have your considered it from that perspective?

Here, DCD and DSR are not being drive by the terminal, these need initializing, so these some  ok.

CTS is *optionally* driven from the terminal DTR line, but it might not be driven at all, it needs initializing, so this also seems ok.

Guess you might argue that initializing the rdx line is not necessary, but why not, it seems prudent.

If there is some state that could be lost here then can anyone point it out? It seems ok, in this instance, unlike some other goof's I've made! The bus interface 'output_*()' functions update the rs232 layer state variables, so they appear ok.